flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/AbstractJobReconciler.java

```diff
@@ -306,6 +319,40 @@ protected void resubmitJob(FlinkResourceContext<CR> ctx, boolean requireHaMetada
         restoreJob(ctx, specToRecover, ctx.getObserveConfig(), requireHaMetadata);
     }
 
+    private void redeployWithSavepoint(
+            FlinkResourceContext<CR> ctx,
+            Configuration deployConfig,
+            CR resource,
+            STATUS status,
+            SPEC currentDeploySpec,
+            JobState desiredJobState)
+            throws Exception {
+        LOG.info("Redeploying from savepoint");
+        cancelJob(ctx, UpgradeMode.STATELESS);
+        var savepointOpt =
+                Optional.ofNullable(currentDeploySpec.getJob().getInitialSavepointPath());
+        currentDeploySpec
+                .getJob()
+                .setUpgradeMode(
+                        savepointOpt.isPresent() ? UpgradeMode.SAVEPOINT : UpgradeMode.STATELESS);
```

Comment:
I feel like we should test the savepoint path is present that before we tear down the job. If it is not there, we can't proceed.

Reply:
if the `initialSavepointPath` is not present the job will be restarted from empty state. At least that's how I implemented and documented it. We could discuss whether this should be allowed or not. The stateless upgradeMode serves this purpose to some extent but then the user has to change the upgradeMode back and forth

Reply:
As I said, I would not expect the savepoint loading to go through if the savepoint path is set.

Reply:
Recommending to use this feature with an empty savepointpath to clear the state, without having to change the upgradeMode to stateless, seems beyond the scope of this change. I don't think we should allow this.

Reply:
makes sense, we can throw a validation error in that case